src/main/java/teammates/ui/webapi/CreateFeedbackQuestionAction.java

```diff
@@ -38,39 +56,166 @@ public JsonResult execute() throws InvalidHttpRequestBodyException {
         String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
 
         FeedbackQuestionCreateRequest request = getAndValidateRequestBody(FeedbackQuestionCreateRequest.class);
-        FeedbackQuestionAttributes attributes = FeedbackQuestionAttributes.builder()
-                .withCourseId(courseId)
-                .withFeedbackSessionName(feedbackSessionName)
-                .withGiverType(request.getGiverType())
-                .withRecipientType(request.getRecipientType())
-                .withQuestionNumber(request.getQuestionNumber())
-                .withNumberOfEntitiesToGiveFeedbackTo(request.getNumberOfEntitiesToGiveFeedbackTo())
-                .withShowResponsesTo(request.getShowResponsesTo())
-                .withShowGiverNameTo(request.getShowGiverNameTo())
-                .withShowRecipientNameTo(request.getShowRecipientNameTo())
-                .withQuestionDetails(request.getQuestionDetails())
-                .withQuestionDescription(request.getQuestionDescription())
-                .build();
 
-        // validate questions (giver & recipient)
-        String err = attributes.getQuestionDetailsCopy().validateGiverRecipientVisibility(attributes);
-        if (!err.isEmpty()) {
-            throw new InvalidHttpRequestBodyException(err);
-        }
-        // validate questions (question details)
-        FeedbackQuestionDetails questionDetails = attributes.getQuestionDetailsCopy();
-        List<String> questionDetailsErrors = questionDetails.validateQuestionDetails();
-        if (!questionDetailsErrors.isEmpty()) {
-            throw new InvalidHttpRequestBodyException(questionDetailsErrors.toString());
+        FeedbackQuestion feedbackQuestion = null;
+        switch (request.getQuestionDetails().getQuestionType()) {
+        case TEXT:
+            feedbackQuestion = new FeedbackTextQuestion(
+                    getNonNullSqlFeedbackSession(feedbackSessionName, courseId),
+                    request.getQuestionNumber(),
+                    request.getQuestionDescription(),
+                    request.getQuestionDetails().getQuestionType(),
+                    request.getGiverType(),
+                    request.getRecipientType(),
+                    request.getNumberOfEntitiesToGiveFeedbackTo(),
+                    request.getShowResponsesTo(),
+                    request.getShowGiverNameTo(),
+                    request.getShowRecipientNameTo(),
+                    request.getQuestionDetails()
```

Comment:
Maybe it's time we consider a factory method for questions and responses. Can have a static method (maybe makeQuestion or similar?) in FeedbackQuestions that takes in the params and the question type and returns the subclass.

Reply:
Sure! Thanks for the advice!